src/gpu/tsvd/tsvd.cu

```diff
@@ -240,12 +239,17 @@ void calculate_u(const Matrix<float> &X, const Matrix<float> &Q, const Matrix<fl
  * @param _param
  */
 void truncated_svd(const double* _X, double* _Q, double* _w, double* _U, double* _explained_variance, double* _explained_variance_ratio, params _param)
+{
+    Matrix<float>X(_param.X_m, _param.X_n);
+	X.copy(_X);
+    truncated_svd(_X, _Q, _w, _U, _explained_variance, _explained_variance_ratio, _param);
+}
+
+void truncated_svd_matrix(Matrix<float> X, double* _Q, double* _w, double* _U, double* _explained_variance, double* _explained_variance_ratio, params _param)
 {
```

Comment:
Why the above change and an additional function? I think tsvd.cu can remain unchanged for this addition?

Reply:
@navdeep-G in PCA I first create the Matrix, then do normalization on it and then pass it to svd, by that time I already have a Matrix object so I would prefer to pass that instead of `double* _X` which would create the X matrix again.

Reply:
Ah, okay, makes sense.

Reply:
@mdymczyk Aren't you only required to center the matrix?  You don't have to normalize it because you aren't creating the covariance matrix -> PCA and instead using SVD.

Reply:
Yes, should be center -> svd, right?

Reply:
FYI, center just means take each column and remove the average so each column average is zero.  Normalization is much more expensive.

Reply:
Yes, exactly. So just need to get each column mean and subtract from each obs. i already have a util to get each column variance. Can take that and make a new one, `mean_iterator`.

Reply:
To get variance. So slight modification for mean. 
```
template<typename T>
class variance_iterator{
public:
    // Required iterator traits
    typedef variance_iterator<T>          self_type;            ///< My own type
    typedef size_t                            difference_type;  ///< Type to express the result of subtracting one iterator from another
    typedef T                           value_type;             ///< The type of the element the iterator can point to
    typedef T*                          pointer;                ///< The type of a pointer to an element the iterator can point to
    typedef T                           reference;              ///< The type of a reference to an element the iterator can point to
    typedef std::random_access_iterator_tag     iterator_category;      ///< The iterator category
	const T* data_ptr;
	const T* mean_ptr;
	const int col_rows;
	size_t offset;
	__device__ T operator[](size_t idx){
		idx = idx + offset;
		T mean = mean_ptr[idx/col_rows];
		T dev_square = pow((data_ptr[idx] - mean),2);
		return dev_square;
	}
	__device__ self_type operator+(size_t idx){
		self_type retval(data_ptr, mean_ptr, col_rows);
		retval.offset += idx;
		return retval;
	}

	__host__ __device__ variance_iterator(const T* data_ptr, const T* mean_ptr, const int col_rows):data_ptr(data_ptr), mean_ptr(mean_ptr), col_rows(col_rows), offset(0){

	}
};
```